src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AbstractRazorDelegatingEndpoint.cs

```diff
@@ -0,0 +1,140 @@
+        /// </summary>
+        public virtual async Task<TResponse?> Handle(TRequest request, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            var documentContext = await _documentContextFactory.TryCreateAsync(request.TextDocument.Uri, cancellationToken).ConfigureAwait(false);
+            if (documentContext is null)
+            {
+                return null;
+            }
+
+            var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken);
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            var sourceText = await documentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+            if (!request.Position.TryGetAbsoluteIndex(sourceText, _logger, out var absoluteIndex))
+            {
+                return null;
+            }
+
+            var projection = await _documentMappingService.GetProjectionAsync(documentContext, absoluteIndex, cancellationToken).ConfigureAwait(false);
+
+            return await GetDelegatedResponseAsync(request, documentContext, codeDocument, sourceText, absoluteIndex, projection, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected async Task<TResponse?> GetDelegatedResponseAsync(
+            TRequest request,
+            DocumentContext documentContext,
+            RazorCodeDocument codeDocument,
+            SourceText sourceText,
+            int absoluteIndex,
+            Projection projection,
+            CancellationToken cancellationToken)
+        {
+            if (!_languageServerFeatureOptions.SingleServerSupport)
+            {
+                return await HandleWithoutSingleServerAsync(request, documentContext, codeDocument, sourceText, cancellationToken).ConfigureAwait(false);
+            }
+
+            if (projection.LanguageKind == RazorLanguageKind.Razor)
+            {
+                return await HandleInRazorAsync(request, documentContext, codeDocument, sourceText, cancellationToken).ConfigureAwait(false);
+            }
+
+            var delegatedParams = CreateDelegatedParams(request, documentContext, projection);
+
+            var delegatedRequest = await _languageServer.SendRequestAsync(EndpointName, delegatedParams).ConfigureAwait(false);
+            var delegatedResponse = await delegatedRequest.Returning<TResponse?>(cancellationToken).ConfigureAwait(false);
+
+            return delegatedResponse is null
+                ? delegatedResponse
+                : await RemapResponseAsync(delegatedResponse, codeDocument, cancellationToken).ConfigureAwait(false);
```

Comment:
I think `ProcessDelegatedResponseAsync` would be a better name. Remapping is a specific type of processing that could be done, but also could be irrelevant.

Also, would it be awesome if the LSP models we used all used a heap of interfaces or generics or something so we could just remap it all here without having to know exactly what the type of the response is? _sigh_ a man can dream. Come on C# 13! :)